### PR TITLE
refactor: move to revm

### PR DESCRIPTION
**Pull request by onbjerg**
Ports our testing backend to REVM. This PR simplifies some code, removes the `evm-adapters` crate, and generally increases testing speed.

More specifically the following parts have been refactored:

- The EVM executor
- Cheatcodes
- Call tracing
- The debugger
- The test runner (some clean up in `ContractRunner`/`MultiContractRunner`)
- Decoding of logs/traces etc.
- Gas reports

The following new features were implemented:

- `forge test --debug` (see https://github.com/gakonst/foundry/pull/920#issuecomment-1068222653)
- `forge run` can now run functions that take arguments (see https://github.com/gakonst/foundry/pull/920#issuecomment-1067344368)
- Fuzzing dictionary see (#985)
- Data from forked endpoints is now optionally cached (see #1006)

The following things were modified:

- The debugger had some minor changes to text, such as titles
- Zero bytes in the stack and memory view are now greyed out

The following bugs were squashed:

- https://github.com/gakonst/foundry/issues/646
- https://github.com/gakonst/foundry/issues/812
- https://github.com/gakonst/foundry/issues/885
- https://github.com/gakonst/foundry/issues/837

The following tests have been added/modified:

- [geb](https://github.com/reflexer-labs/geb) was re-enabled in our integration tests
- [Lil' Web3](https://github.com/m1guelpf/lil-web3/) was added to our integration tests
- [Maple Labs' Loan](https://github.com/maple-labs/loan/) was added to our integration tests
- We now have a common set of Solidity tests we can test against. The idea is that we expand this suite to include every aspect of Foundry that touches Solidity files, including the linter/formatter and new test types (see https://github.com/gakonst/foundry/pull/965)

## Tasks

Blocking issues:

- [x] https://github.com/gakonst/foundry/issues/753 (#867)
- [x] https://github.com/gakonst/foundry/issues/775 (#867)
- [x] https://github.com/gakonst/foundry/issues/758 (#920)
- [x] https://github.com/gakonst/foundry/issues/761 (#920)
- [x] Tracking issue for new features that need to be ported: https://github.com/gakonst/foundry/issues/799

Things that would be nice to address:

- [x] https://github.com/gakonst/foundry/issues/812 (might already be fixed)
- [ ] https://github.com/gakonst/foundry/issues/654
- [ ] https://github.com/gakonst/foundry/issues/876

Things that would be nice to add:

- [ ] 4byte trace identifier
- [ ] `ethers-addressbook` trace identifier (#640)
- [ ] Etherscan trace identifier (https://github.com/gakonst/foundry/issues/447)
- [x] Common Solidity tests in root of project: currently some crates duplicate tests or reference tests in other crates (https://github.com/onbjerg/foundry-test)
- [x] `forge test --debug`

Checklist before merge:

- [x] Compare gas usage between the port and master
- [x] Compare trace decoding between the port and master
- [x] Link all fixes issues
- [x] Create issues in the Foundry Book repository with updates
- [x] Create issues for unresolved stretch tasks

Everything in the PR has been tested against https://github.com/onbjerg/foundry-test and solmate

### CODE REVIEW

**Comment by gakonst** [comment]
I think the Foundry-Test repo would be nice to get in the main PR and have MultiContract runner run against it, as it gets us a lot of confidence about the safety of the code. Otherwise agree that Etherscan/4Byte can happen in separate PR, as long as we believe that the current architecture is generic enough to support batching most of the RPC requests for signatures/abis etc so we don't have big performance overheads

**Comment by onbjerg** [comment]
Current rough benchmarks. For reference, solmate is currently 186 fuzz tests and 209 unit tests

**Tracing off (256 fuzz runs)**
image

**Full tracing on (256 fuzz runs)**
image

**Comment by gakonst** [comment]
Turns out compiling with `RUSTFLAGS="-C target-cpu=native" cargo build --release --bin forge` yields non-trivial speed increase, on the order of ~10%

**Comment by onbjerg** [comment]
Ok so this is pretty much good to go with the exception of a memory allocation bug in REVM that I will address tomorrow. Gonna take tomorrow to work on this for the final time before merge, will try to squeeze in some optimizations in REVM and #731 / #849  / an Etherscan decoder if possible

**Comment by onbjerg** [comment]
Going to fix the failing tests, otherwise this is good to go

**Comment by onbjerg** [comment]
Just noticed `SSTORE2Test::testFailReadInvalidPointer(address)` in Solmate will correctly revert but still be marked as failed. Investigating

File: evm/src/executor/abi.rs [hunk 1]
```diff
@@ -1,53 +1,15 @@
...
-        let slot = account.storage.entry(slot).or_insert_with(Default::default);
-        *slot = value;
-    }
-}
+/// The cheatcode handler address.
+///
+/// This is the same address as the one used in DappTools's HEVM.
+pub static CHEATCODE_ADDRESS: Lazy<Address> = Lazy::new(|| {
+    Address::from_slice(&hex::decode("7109709ECfa91a80626fF3989D68f67F5b1DD12D").unwrap())
+});
```
Comment (mattsse) [inline evm/src/executor/abi.rs:10]: this could be made a const/static, no need for this to be a Lazy

File: evm/src/executor/abi.rs [hunk 2]
```diff
@@ -81,8 +44,16 @@ ethers::contract::abigen!(
 );
 pub use hevm_mod::{HEVMCalls, HEVM_ABI};
 
+/// The Hardhat console address.
+///
+/// See: https://github.com/nomiclabs/hardhat/blob/master/packages/hardhat-core/console.sol
+pub static HARDHAT_CONSOLE_ADDRESS: Lazy<Address> = Lazy::new(|| {
+    Address::from_slice(&hex::decode("000000000000000000636F6e736F6c652e6c6f67").unwrap())
+});
```
Comment (mattsse) [inline evm/src/executor/abi.rs:52]: same, could initialize the raw H160 as static

File: evm/src/executor/inspector/cheatcodes/expect.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+/// Solidity will see a successful call and attempt to decode the return data. Therefore, we need
+/// to populate the return with dummy bytes so the decode doesn't fail.
+///
+/// 320 bytes was arbitrarily chosen because it is long enough for return values up to 10 words in
+/// size.
+static DUMMY_CALL_OUTPUT: [u8; 320] = [0u8; 320];
+
+/// Same reasoning as [DUMMY_CALL_OUTPUT], but for creates.
+static DUMMY_CREATE_ADDRESS: Lazy<Address> =
+    Lazy::new(|| Address::from_str("0000000000000000000000000000000000000001").unwrap());
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/expect.rs:22]: same,

File: evm/src/executor/inspector/stack.rs [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+            if let Some($id) = $inspector {
+                $call;
+            }
+        })+
+    }
+}
+
+/// An inspector that calls multiple inspectors in sequence.
+///
+/// The order in which inspectors are called match the order in which they were added to the stack.
```
Comment (mattsse) [inline evm/src/executor/inspector/stack.rs:20]: this is no longer up to date, right? as these are now simple fields

File: evm/src/executor/inspector/stack.rs [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+    pub logs: Option<LogCollector>,
+    pub cheatcodes: Option<Cheatcodes>,
+    pub debugger: Option<Debugger>,
+}
+
+impl InspectorStack {
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
```
Comment (mattsse) [inline evm/src/executor/inspector/stack.rs:36]: recommend removing
```suggestion
```

File: evm/src/executor/inspector/stack.rs [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+where
+    DB: Database,
+{
+    fn initialize_interp(
+        &mut self,
+        interpreter: &mut Interpreter,
+        data: &mut EVMData<'_, DB>,
+        is_static: bool,
+    ) -> Return {
+        call_inspectors!(
```
Comment (mattsse) [inline evm/src/executor/inspector/stack.rs:48]: nit: seems like this macro is used a bunch,

I think this always calls the same inspectors, right? followed by a closure

can we instead call all inspectors in a single function instead?

File: evm/src/executor/inspector/tracer.rs [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+#[derive(Default, Debug)]
+pub struct Tracer {
+    pub trace_stack: Vec<usize>,
+    pub traces: CallTraceArena,
+}
+
+impl Tracer {
+    pub fn new() -> Self {
+        Default::default()
+    }
```
Comment (mattsse) [inline evm/src/executor/inspector/tracer.rs:28]: remove?

File: evm/src/executor/mod.rs [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+}
+
+struct InspectorData {
+    logs: Vec<RawLog>,
+    labels: BTreeMap<Address, String>,
+    traces: Option<CallTraceArena>,
+    debug: Option<DebugArena>,
+}
+
+fn collect_inspector_states(stack: InspectorStack) -> InspectorData {
```
Comment (mattsse) [inline evm/src/executor/mod.rs:496]: make this `InspectorStack::collect_inspector_states` instead

File: evm/src/fuzz/mod.rs [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+use revm::db::DatabaseRef;
+use serde::{Deserialize, Serialize};
+use std::{cell::RefCell, collections::BTreeMap, fmt};
+use strategies::{
+    build_initial_state, collect_state_from_call, fuzz_calldata, fuzz_calldata_from_state,
+    EvmFuzzState,
+};
+
+/// Magic return code for the `assume` cheatcode
+pub const ASSUME_MAGIC_RETURN_CODE: &[u8] = "FOUNDRY::ASSUME".as_bytes();
```
Comment (mattsse) [inline evm/src/fuzz/mod.rs:23]:
```suggestion
pub const ASSUME_MAGIC_RETURN_CODE: &[u8] = b"FOUNDRY::ASSUME";
```

**Review by mattsse — CHANGES_REQUESTED** [review]
couple nits, not necessarily introduced in this PR.

we should consider cleaning up some Test/TestRunner related code but would recommend doing this after this is merged.

File: evm/src/executor/inspector/stack.rs (hunk 6 above)
Comment (onbjerg) [reply evm/src/executor/inspector/stack.rs:48]: Don't think so, the function that is being called on every inspector changes every time. Do you have any suggestions?

Comment (mattsse) [reply evm/src/executor/inspector/stack.rs:63]: oh I see the call is invoked for each inspector: `$call;`

and we want to be able to return directly, so we need to expand the code in place, makes sense.

**Review by mattsse — APPROVED** [review]
can't spot anything obvious anymore.
let's fix the things we missed once they come in

**Review by gakonst — APPROVED** [review]
Ok let's do it - n

**Pull request merged**